pkg/tcpip/stack/conntrack.go

```diff
@@ -0,0 +1,253 @@
+	src connTrackManip
+
+	// dst is non-manipulatable part of the tuple.
+	dst connTrackNonManip
+}
+
+// connTrackTupleHash is the container of tuple and connection.
+type connTrackTupleHash struct {
+	// ct is pointer to the connection tracking entry.
+	ct *connTrack
+
+	// tuple is original or reply tuple.
+	tuple connTrackTuple
+}
+
+// connTrack is the connection.
+type connTrack struct {
+	// tupleHash contains two tuple hashes one for each direction.`
+	tupleHash [CT_DIR_MAX]connTrackTupleHash
+
+	// status indicates connection is new or established.
+	status uint32
+
+	// timeout indicates the time connection should be active.
+	timeout uint32
+}
+
+// packetToTuple converts packet to tuple.
+func packetToTuple(pkt PacketBuffer, hook Hook) (connTrackTuple, *tcpip.Error) {
+	var tuple connTrackTuple
+	var tcpHeader header.TCP
+	var netHeader header.IPv4
+
+	// Parse packet to get tuple.
+	// If the hook is prerouting, headers should be set.
+	// TODO(gvisor.dev/issue/170): Change this after parsing headers code
+	// is added.
+	if hook == Prerouting {
+		newPkt := pkt.Clone()
+
+		// Set network header.
+		headerView := newPkt.Data.First()
+		netHeader = header.IPv4(headerView)
+		newPkt.NetworkHeader = headerView[:header.IPv4MinimumSize]
+
+		hlen := int(netHeader.HeaderLength())
+		tlen := int(netHeader.TotalLength())
+		newPkt.Data.TrimFront(hlen)
+		newPkt.Data.CapLength(tlen - hlen)
+
+		// TODO(gvisor.dev/issue/170): Need to support for other
+		// protocols as well.
+		protocol := netHeader.TransportProtocol()
+		if protocol != header.TCPProtocolNumber {
+			return tuple, tcpip.ErrUnknownProtocol
+		}
+
+		// Set transport header.
+		if newPkt.TransportHeader != nil {
+			tcpHeader = header.TCP(newPkt.TransportHeader)
+		} else {
+			if len(pkt.Data.First()) < header.TCPMinimumSize {
+				return tuple, tcpip.ErrUnknownProtocol
+			}
+			tcpHeader = header.TCP(newPkt.Data.First())
+		}
+	} else {
+		netHeader = header.IPv4(pkt.NetworkHeader)
+		// TODO(gvisor.dev/issue/170): Need to support for other
+		// protocols as well.
+		if netHeader == nil || netHeader.TransportProtocol() != header.TCPProtocolNumber {
+			return tuple, tcpip.ErrUnknownProtocol
+		}
+		tcpHeader = header.TCP(pkt.TransportHeader)
+	}
+
+	tuple.src.addr = netHeader.SourceAddress()
+	tuple.src.port = tcpHeader.SourcePort()
+	tuple.src.protocol = uint16(header.IPv4ProtocolNumber)
+
+	tuple.dst.addr = netHeader.DestinationAddress()
+	tuple.dst.port = tcpHeader.DestinationPort()
+
+	// TODO(gvisor.dev/issue/170): Need to support other transport protocols.
+	tuple.dst.protocol = uint8(header.TCPProtocolNumber)
+	tuple.dst.direction = CT_DIR_ORIGINAL
+
+	return tuple, nil
+}
+
+// getConnTrack returns connection if exists and sets status of connection
+// to established if the tuple is reply of an existing connection.
+func getConnTrack(hash uint32, s *Stack) (*connTrack, *tcpip.Error) {
+	tupleHash, err := s.connTrackTable[hash]
```

Comment:
s/err/ok/. `ok` is the idiomatic Go term for this sort of bool.

Similarly, `getConnTrack` can probably return `(*connTrack, bool)` and be handled the same way.

Reply:
Done.